src/Bicep.LangServer/BicepCompilationManager.cs

```diff
@@ -17,72 +21,146 @@ namespace Bicep.LanguageServer
-                        {
-                            Start = new Position(0, 0),
-                            End = new Position(1, 0),
-                        },
-                        Severity = DiagnosticSeverity.Error,
-                        Message = exception.Message,
-                        Code = new DiagnosticCode("Fatal")
-                    }
-                });
-
-                return null;
+                        Start = new Position(0, 0),
+                        End = new Position(1, 0),
+                    },
+                    Severity = DiagnosticSeverity.Error,
+                    Message = exception.Message,
+                    Code = new DiagnosticCode("Fatal")
+                };
+                
+                // the file is no longer in a state that can be parsed
+                // clear all info to prevent cascading failures elsewhere
+                var closedTrees = CloseCompilationInternal(documentUri, version, fatalError.AsEnumerable());
+
+                return (ImmutableArray<SyntaxTree>.Empty, closedTrees);
             }
         }
 
-        public void CloseCompilation(DocumentUri uri)
+        public void UpsertCompilation(DocumentUri uri, int? version, string fileContents)
         {
-            // remove the active compilation
-            this.activeContexts.TryRemove(uri, out _);
+            var newSyntaxTree = this.provider.BuildSyntaxTree(uri, fileContents);
+            var firstChanges = workspace.UpsertSyntaxTrees(newSyntaxTree.AsEnumerable());
+            var secondChanges = UpdateCompilationInternal(uri, version);
+            
+            var addedTrees = firstChanges.added.Concat(secondChanges.added);
+            var removedTrees = firstChanges.removed.Concat(secondChanges.removed);
 
-            // clear diagnostics for the file
+            foreach (var (entrypointUri, context) in activeContexts)
+            {
+                if (removedTrees.Any(x => context.Compilation.SyntaxTreeGrouping.SyntaxTrees.Contains(x)))
+                {
+                    UpdateCompilationInternal(entrypointUri, null);
+                }
+            }
+        }
+
+        public void CloseCompilation(DocumentUri documentUri)
+        {
+            // close and clear diagnostics for the file
             // if upsert failed to create a compilation due to a fatal error, we still need to clean up the diagnostics
-            PublishDocumentDiagnostics(uri, 0, Enumerable.Empty<OmniSharp.Extensions.LanguageServer.Protocol.Models.Diagnostic>());
+            CloseCompilationInternal(documentUri, 0, Enumerable.Empty<OmniSharp.Extensions.LanguageServer.Protocol.Models.Diagnostic>());
+        }
+
+        public void HandleFileChanges(IEnumerable<FileEvent> fileEvents)
+        {
+            var removedTrees = new HashSet<SyntaxTree>();
+            foreach (var change in fileEvents.Where(x => x.Type == FileChangeType.Changed || x.Type == FileChangeType.Deleted))
```

Comment:
I'm not handling the case where a non-existent file which is being referenced as a module is created in the background. Probably should, but it'll involve caching negative SyntaxTree lookups as well.